mir/src/passes/unrolling/match_optimizer.rs

```diff
@@ -0,0 +1,241 @@
+        let mut bus_related_constraints_for_match_arm = Vec::new();
+
+        // 1.2. Filter out all BusOp nodes from the constraints, we will handle them separately
+        for constraint in all_constraints {
+            match constraint.borrow().deref() {
+                Op::BusOp(_) => bus_related_constraints_for_match_arm.push(constraint.clone()),
+                Op::Enf(enf) => match enf.expr.borrow().deref() {
+                    Op::BusOp(_) => bus_related_constraints_for_match_arm.push(enf.expr.clone()),
+                    _ => constraints_to_eval.push(constraint.clone()),
+                },
+                _ => constraints_to_eval.push(constraint.clone()),
+            }
+        }
+
+        // 1.3. Evaluate all the other constraints at random points
+        for constraint in constraints_to_eval {
+            let eval = self.random_inputs.eval(constraint.clone())?;
+            // Check if we already have this eval in our list
+            match self.node_evals.iter().position(|e| e == &eval) {
+                Some(index) => {
+                    // If the eval is already in the list, we use its index in the
+                    // node_evals vec
+                    let Some(constraints_vec) = self.constraints_evaluation_indices.get_mut(&index)
+                    else {
+                        unreachable!(
+                            "Error: Reference to an evaluation index not found in constraints_evaluation_indices"
+                        );
+                    };
+                    // We add the constraint to the corresponding vector only if the
+                    // condition is not already present
+                    // to remove duplicate constraints for the same selector
+                    if !constraints_vec.iter().any(|(c, _)| c == &condition) {
+                        constraints_vec.push((condition.clone(), constraint.clone()))
+                    }
+                },
+                None => {
+                    // Otherwise, we add it to the list and use its index
+                    self.node_evals.push(eval);
+                    let index = self.node_evals.len() - 1;
+                    if self
+                        .constraints_evaluation_indices
+                        .insert(index, vec![(condition.clone(), constraint.clone())])
+                        .is_some()
+                    {
+                        unreachable!(
+                            "Error: A new evaluation index was found in constraints_evaluation_indices"
+                        );
+                    }
+                },
+            }
+        }
+        Ok(bus_related_constraints_for_match_arm)
+    }
+
+    /// For each evaluation, computes the number of constraints (with different selectors)
+    /// evaluating to it.
+    // Returns eval_lens: BTreeMap<len, Vec<evaluation_index>>
+    pub fn compute_eval_lens(&self) -> BTreeMap<usize, Vec<usize>> {
```

Comment:
A couple of nits: 
- this could be just a helper function, right? (i.e., not `pub`)
- Should it be `///` online 117 rather than `//`?